### PR TITLE
Unit tests for Weighted Arithmetic Mean and Distribution

### PR DESCRIPTION
**Pull request by nathannaveen**
- Added unit tests for WAM.

Signed-off-by: nathannaveen <42319948+nathannaveen@users.noreply.github.com>

### CODE REVIEW

File: internal/scorer/algorithm/wam/wam_test.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+				inputs: []*algorithm.Input{
+					{Weight: 0, Distribution: algorithm.LookupDistribution("linear"),
+						Source: algorithm.Value(algorithm.Field("0"))},
+					{Weight: 1, Distribution: algorithm.LookupDistribution("linear"),
+						Source: algorithm.Value(algorithm.Field("1"))},
+				},
+			},
+			want: 1,
+			args: args{
+				record: map[string]float64{"1": 1},
```
Comment (calebbrown) [inline internal/scorer/algorithm/wam/wam_test.go:70]: This should have a field for "0" so that it is clear the weight is excluding the value.

**Review by calebbrown — CHANGES_REQUESTED** [review]
Thanks for implementing these tests!

File: internal/scorer/algorithm/wam/wam.go [hunk 2]
```diff
@@ -15,6 +15,7 @@
 // The package wam implements the Weighted Arithmetic Mean, which forms the
 // basis of Rob Pike's criticality score algorithm as documented in
 // Quantifying_criticality_algorithm.pdf.
+
```
[the remark this answers is not in the document]
Comment (nathannaveen) [reply internal/scorer/algorithm/wam/wam.go]: Removed!

File: internal/scorer/algorithm/distribution_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+// limitations under the License.
+
+package algorithm
+
+import (
+	"math"
+	"testing"
+)
+
+func TestLookupDistribution(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (nathannaveen) [reply internal/scorer/algorithm/distribution_test.go:22]: Fixed

File: internal/scorer/algorithm/wam/wam_test.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+	type fields struct {
+		inputs []*algorithm.Input
+	}
+	type args struct {
+		record map[string]float64
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
```
[the remark this answers is not in the document]
Comment (nathannaveen) [reply internal/scorer/algorithm/wam/wam_test.go]: I took care of this.

File: internal/scorer/algorithm/wam/wam_test.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+						Source: algorithm.Value(algorithm.Field("2"))},
+					{Weight: 3, Distribution: algorithm.LookupDistribution("linear"),
+						Source: algorithm.Value(algorithm.Field("3"))},
+					{Weight: 4, Distribution: algorithm.LookupDistribution("linear"),
+						Source: algorithm.Value(algorithm.Field("4"))},
+					{Weight: 5, Distribution: algorithm.LookupDistribution("linear"),
+						Source: algorithm.Value(algorithm.Field("5"))},
+				},
+			},
+			want: 3.6666666666666665,
```
[the remark this answers is not in the document]
Comment (nathannaveen) [reply internal/scorer/algorithm/wam/wam_test.go]: The code itself returns 3.66... so that is why I have the test wanting 3.66. What are your thoughts?

File: internal/scorer/algorithm/wam/wam_test.go (hunk 1 above)
Comment (nathannaveen) [reply internal/scorer/algorithm/wam/wam_test.go]: I have changed that only to have a field of 0.

File: internal/scorer/algorithm/wam/wam_test.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+						Source: algorithm.Value(algorithm.Field("0"))},
+				},
+			},
+			want: math.NaN(),
+			args: args{
+				record: map[string]float64{"1": 1},
+			},
+		},
+	}
+	for _, tt := range tests {
```
[the remark this answers is not in the document]
Comment (nathannaveen) [reply internal/scorer/algorithm/wam/wam_test.go]: 👍

File: internal/scorer/algorithm/distribution_test.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+			distributionName:  "zipfian",
+			distributionValue: math.MaxInt64,
+
+			want: result{
+				name:  "zipfian",
+				value: 43.668272,
+			},
+		},
+	}
+	for _, tt := range tests {
```
Comment (calebbrown) [inline internal/scorer/algorithm/distribution_test.go:129]: Please use `test` instead of `tt`

File: internal/scorer/algorithm/distribution_test.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+			}
+
+			normalizedValue := got.Normalize(tt.distributionValue)
+
+			// round to 6 decimal places for comparison
+			normalizedValue = math.Round(normalizedValue*1000000) / 1000000
+			wantValue := math.Round(tt.want.value*1000000) / 1000000
+
+			if got.String() != tt.want.name {
+				t.Errorf("The names are not equal LookupDistribution name = %s, want name %s",
```
Comment (calebbrown) [inline internal/scorer/algorithm/distribution_test.go:144]: nit: the text "The names are not equal" is not necessary, the rest of the string and the presence of the error makes the inequality obvious.

File: internal/scorer/algorithm/distribution_test.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+					got.String(), tt.want.name)
+			}
+
+			if math.IsNaN(normalizedValue) && math.IsNaN(tt.want.value) {
+				// both are NaN, and we can't compare two NaNs together
+				return
+			}
+
+			if normalizedValue != wantValue {
+				t.Errorf("The values are not equal LookupDistribution value %f, want value %f",
```
Comment (calebbrown) [inline internal/scorer/algorithm/distribution_test.go:154]: nit: as above, the text "The values are not equal" is not necessary.

**Review by calebbrown — CHANGES_REQUESTED** [review]

File: internal/scorer/algorithm/wam/wam_test.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+				{
+					Weight: 4, Distribution: algorithm.LookupDistribution("linear"),
+					Source: algorithm.Value(algorithm.Field("4")),
+				},
+				{
+					Weight: 5, Distribution: algorithm.LookupDistribution("linear"),
+					Source: algorithm.Value(algorithm.Field("5")),
+				},
+			},
+			want:   3.6666666666666665,
```
[the remark this answers is not in the document]
Comment (nathannaveen) [reply internal/scorer/algorithm/wam/wam_test.go:57]: I realized why I got `3.666` and you got `1`, I ran the tests with `record: map[string]float64{"1": 1, "2": 2, "3": 3, "4": 4, "5": 5}`, but forgot to commit that so it was only `record: map[string]float64{"1": 1},`. I have fixed this, so the record is `record: map[string]float64{"1": 1, "2": 2, "3": 3, "4": 4, "5": 5}`.

File: internal/scorer/algorithm/wam/wam_test.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+			record: map[string]float64{"1": 1},
+		},
+		{
+			name: "With zero weight",
+			inputs: []*algorithm.Input{
+				{
+					Weight: 0, Distribution: algorithm.LookupDistribution("linear"),
```
[the remark this answers is not in the document]
Comment (nathannaveen) [reply internal/scorer/algorithm/wam/wam_test.go:64]: I updated `LoadConfig()` from `config.go`, and `New()` from `wam.go` with the check for the weight being greater than `0`.

File: internal/scorer/algorithm/wam/wam_test.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+			},
+			want:   1,
+			record: map[string]float64{"1": 1},
+		},
+		{
+			name: "with a single negative weight",
+			inputs: []*algorithm.Input{
+				{
+					Weight: -300, Distribution: algorithm.LookupDistribution("linear"),
+					Source: algorithm.Value(algorithm.Field("-300")),
```
[the remark this answers is not in the document]
Comment (nathannaveen) [reply internal/scorer/algorithm/wam/wam_test.go:91]: I fixed this by updating the negative weights.

File: internal/scorer/algorithm/distribution_test.go [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+
+			if got == nil && tt.wantNil {
+				return
+			}
+
+			normalizedValue := got.Normalize(tt.distributionValue)
+
+			// round to 6 decimal places for comparison
+			normalizedValue = math.Round(normalizedValue*1000000) / 1000000
+			wantValue := math.Round(tt.want.value*1000000) / 1000000
```
[the remark this answers is not in the document]
Comment (nathannaveen) [reply internal/scorer/algorithm/distribution_test.go]: Agreed! I have taken care of this!

File: internal/scorer/algorithm/distribution_test.go (hunk 7 above)
Comment (nathannaveen) [reply internal/scorer/algorithm/distribution_test.go]: Done!

File: internal/scorer/algorithm/distribution_test.go (hunk 8 above)
Comment (nathannaveen) [reply internal/scorer/algorithm/distribution_test.go]: 👍

File: internal/scorer/algorithm/distribution_test.go (hunk 9 above)
Comment (nathannaveen) [reply internal/scorer/algorithm/distribution_test.go]: 👍

File: internal/scorer/algorithm/wam/wam.go [hunk 14]
```diff
@@ -18,16 +18,26 @@
...
 
 // New returns a new instance of the Weighted Arithmetic Mean algorithm, which
 // is used by the Pike algorithm.
 func New(inputs []*algorithm.Input) (algorithm.Algorithm, error) {
+	for _, i := range inputs {
+		if i.Weight <= 0 {
+			return nil, ErrWeightLessThanOrEqualToZero
+		}
+	}
+
```
Comment (calebbrown) [inline internal/scorer/algorithm/wam/wam.go:40]: This check belongs better in `config.go` or a new function for constructing Input instances.

File: internal/scorer/algorithm/wam/wam_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (calebbrown) [reply internal/scorer/algorithm/wam/wam_test.go:57]: Thanks for fixing. That makes more sense.

File: internal/scorer/algorithm/wam/wam_test.go [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+				},
+				{
+					Weight: 5, Distribution: algorithm.LookupDistribution("linear"),
+					Source: algorithm.Value(algorithm.Field("5")),
+				},
+			},
+			want:   3.6666666666666665,
+			record: map[string]float64{"1": 1, "2": 2, "3": 3, "4": 4, "5": 5},
+		},
+		{
```
Comment (calebbrown) [inline internal/scorer/algorithm/wam/wam_test.go:60]: If these are blocked when the Input is created then I don't think these "negative weight" or "zero weight" tests need to be here.

However there should be tests here for edge cases, for example:

- no fields matching the record
- partial matches between the defined fields and the record

**Review by calebbrown — CHANGES_REQUESTED** [review]

File: internal/scorer/algorithm/wam/wam.go (hunk 14 above)
Comment (nathannaveen) [reply internal/scorer/algorithm/wam/wam.go:40]: I don't understand what you mean, could you please abbreviate?

Comment (calebbrown) [reply internal/scorer/algorithm/wam/wam.go:40]: Sure.

Currently, this check will only be applied to the Weighted Arithmetic Mean.

The check should be applied across all possible algorithms as zero weights are no-ops and negative weights are confusing and error prone.

The most sensible place to have this check is inside the config.go in `UnmarshalYAML()` while the Input is being parsed.

However, another place it could be added is in `algorithm/input.go` by creating a 'NewInput()` method that does the validation which would be called from `ToAlgorithmInput()` in config.go.

File: internal/scorer/config.go [hunk 16]
```diff
@@ -121,6 +123,13 @@ func LoadConfig(r io.Reader) (*Config, error) {
 	if err := yaml.Unmarshal(data, c); err != nil {
 		return nil, err
 	}
+
+	for _, i := range c.Inputs {
+		if i.Weight <= 0 {
+			return nil, wam.ErrWeightLessThanOrEqualToZero
+		}
```
Comment (calebbrown) [inline internal/scorer/config.go:130]: See my comment re:this error check.

Additionally, the error `ErrWeightLessThanOrEqualToZero` should not be pulled from `wam`.

**Review by calebbrown — CHANGES_REQUESTED** [review]

File: internal/scorer/algorithm/wam/wam.go (hunk 14 above)
Comment (nathannaveen) [reply internal/scorer/algorithm/wam/wam.go]: Thank you for the explanation. I have added the check into `UnmarshalYAML()`.

File: internal/scorer/algorithm/wam/wam_test.go (hunk 15 above)
Comment (nathannaveen) [reply internal/scorer/algorithm/wam/wam_test.go:61]: I have removed the negative weight and zero weight tests, and have put in the tests `"fields not matching the record"` and `"some fields matching the record"`.

File: internal/scorer/config.go (hunk 16 above)
Comment (nathannaveen) [reply internal/scorer/config.go]: 👍

File: internal/scorer/config.go [hunk 17]
```diff
@@ -52,6 +52,11 @@ func (i *Input) UnmarshalYAML(value *yaml.Node) error {
 		return errors.New("field must be set")
 	}
 	*i = Input(*raw)
+
+	if i.Weight <= 0 {
+		return errors.New("weight must be greater than 0")
+	}
+
```
Comment (calebbrown) [inline internal/scorer/config.go:59]: nit: please but this under the `raw.Field` check and use `raw.Weight` instead of `i.Weight`

Otherwise this is exactly what I had in mind! thanks :)

**Review by calebbrown — CHANGES_REQUESTED** [review]
Looking good, just one minor change!

File: internal/scorer/config.go (hunk 17 above)
Comment (nathannaveen) [reply internal/scorer/config.go]: Took care of it, thanks

**Review by calebbrown — APPROVED** [review]
Thanks!

**Review by calebbrown — APPROVED** [review]

**Comment by nathannaveen** [comment]
> Thanks!

Yay! Thanks for all your reviews! Merged! 🥳
